src/autora/experimentalist/random_.py

```diff
@@ -0,0 +1,341 @@
+
+
+@singledispatch
+def random_pool(s, **kwargs):
+    """Function to create a sequence of conditions randomly sampled from independent variables."""
+    raise NotImplementedError(
+        "random_pool doesn't have an implementation for %s (type=%s)" % (s, type(s))
+    )
+
+
+@random_pool.register(State)
+def random_pool_on_state(
+    s: State,
+    num_samples: int = 5,
+    random_state: Optional[int] = None,
+    replace: bool = True,
+) -> State:
+    """
+
+    Args:
+        variables:
+        fmt: the output type required
+
+    Returns:
+
+    Examples:
+        >>> from autora.state.delta import State
+        >>> from autora.variable import VariableCollection, Variable
+        >>> from dataclasses import dataclass, field
+        >>> import pandas as pd
+        >>> import numpy as np
+
+        We define a state object with the fields we need:
+        >>> @dataclass(frozen=True)
+        ... class S(State):
+        ...     variables: VariableCollection = field(default_factory=VariableCollection)
+        ...     conditions: pd.DataFrame = field(default_factory=pd.DataFrame,
+        ...                                      metadata={"delta": "replace"})
+
+        With one independent variable "x", and some allowed_values:
+        >>> s = S(
+        ...     variables=VariableCollection(independent_variables=[
+        ...         Variable(name="x", allowed_values=range(10))
+        ... ]))
+
+        ... we get some of those values back when running the experimentalist:
+        >>> random_pool(s, random_state=1).conditions
+           x
+        0  4
+        1  5
+        2  7
+        3  9
+        4  0
+
+        With one independent variable "x", and a value_range:
+        >>> t = S(
+        ...     variables=VariableCollection(independent_variables=[
+        ...         Variable(name="x", value_range=(-5, 5))
+        ... ]))
+
+        ... we get a sample of the range back when running the experimentalist:
+        >>> random_pool(t, random_state=1).conditions
+                  x
+        0  0.118216
+        1  4.504637
+        2 -3.558404
+        3  4.486494
+        4 -1.881685
+
+
+
+        The allowed_values or value_range must be specified:
+        >>> random_pool(
+        ...     S(variables=VariableCollection(independent_variables=[Variable(name="x")])))
+        Traceback (most recent call last):
+        ...
+        ValueError: allowed_values or [value_range and type==REAL] needs to be set...
+
+        With two independent variables, we get independent samples on both axes:
+        >>> t = S(
+        ...     variables=VariableCollection(independent_variables=[
+        ...         Variable(name="x1", allowed_values=range(1, 5)),
+        ...         Variable(name="x2", allowed_values=range(1, 500)),
+        ... ]))
+        >>> random_pool(t,
+        ...                            num_samples=10, replace=True, random_state=1).conditions
```

Comment:
does having this bit up on the previous line just make it too long for our linter?

Reply:
No, previously the function name was much longer. The formatting is fixed in the next update.